src/Sparrow.Server/Allocators.PointerTypes.cs

```diff
@@ -0,0 +1,527 @@
+        {
+            if (length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+
+            return new Span<T>(Address, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Span<T> AsSpan(int start, int length)
+        {
+            if (start + length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+            
+            return new Span<T>((byte*)Address + start * Unsafe.SizeOf<T>(), length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpan<byte> AsReadOnlySpan()
+        {
+            return new ReadOnlySpan<byte>(Address, (int)Size);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpan<T> AsReadOnlySpan(int length)
+        {
+            if (length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+
+            return new ReadOnlySpan<T>(Address, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpan<T> AsReadOnlySpan(int start, int length)
+        {
+            if (start + length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+
+            return new ReadOnlySpan<T>((byte*)Address + start * Unsafe.SizeOf<T>(), length);
+        }
+
+        public int Size
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                EnsureIsNotBadPointerAccess();
+                // We move to the actual header which is behind.
+                return _size;
+            }
+        }
+
+        public int SizeAsBytes
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                EnsureIsNotBadPointerAccess();
+                // We move to the actual header which is behind.
+                return _size * Unsafe.SizeOf<T>();
+            }
+        }
+
+        [Conditional("DEBUG")]
+        [Conditional("VALIDATE")]
+        internal void EnsureIsNotBadPointerAccess()
+        {
+            if (Address == null)
+                throw new InvalidOperationException($"Trying to access the pointer but it is not valid");
+        }
+
+        public string Describe()
+        {
+            if (this.IsValid)
+                return $"{{{(long)this.Address:X64}|{this.Size}|{this.BlockSize}|{this.SizeAsBytes}b}}";
+
+            return "{null}";
+        }
+
+        public override string ToString()
+        {
+            return Describe();
+        }
+
+        public static implicit operator BlockPointer(BlockPointer<T> ptr)
+        {
+            return new BlockPointer(ptr.Address, ptr.BlockSize, ptr.Size);
```

Comment:
```suggestion
            return new BlockPointer(ptr.Address, ptr.BlockSize, ptr.SizeAsBytes);
```

Reply:
And I'm not sure about the `BlockSize` here either, is that in bytes, in elements?

Reply:
Actually, it looks like I've never used/tested those conversions... most of them are wrong.